### ISSUE TITLE
Fl_Counter slips into infinite loop

### ISSUE THREAD

**Issue by ramcdona** [fltk/fltk#450]
As discussed on the fltk.coredev mailing list [here]( https://groups.google.com/g/fltkcoredev/c/daKlBxeOJVk/m/D-MPHPiCAAAJ), Fl_Counter will sometimes slip into an infinite loop when incremented.

This issue was introduced by commit [29d9e31c51e6c](https://github.com/fltk/fltk/commit/29d9e31c51e6c).

It was first noticed on MacOS, but has been confirmed on Linux as well.  Albrecht supplied a test example in the mailing list thread.

The issue does not occur if the user's application finishes handling the event quickly.  However, if there is some delay, it will continue triggering events.  Albrecht's test was able to use a call to usleep() to trigger the problem reproducibly.

**Comment by Albrecht-S** [comment]
After some research I found out that there are two (or three) factors that combine their effects:

1. The new timeout handling could cause "new" timeouts, i.e. those added inside another button or timer callback to be executed immediately w/o intervening event handling if the timer had already expired. This could happen if `Fl::repeat_timeout()` was called inside a callback whose handling took longer than the requested timeout. Under certain circumstances the same issue could be caused by adding another timeout with `Fl::add_timeout()`. The Fl_Counter widget does this.

Commit f3bf231cc06ab82576b798a1f615981924a2903c fixes this issue.

2. The event handling of Fl_Counter could miss some events, e.g. if a modal window was created in the callback of Fl_Counter and maybe under some other circumstances.

Commit 232743c3a5d903be813f6c4445f3f96bab25cae0 fixes this issue by checking the focus and mouse button states.

3. Further testing on macOS revealed some "special effects" that could not be reproduced on Linux and Windows. As the OP @ramcdona described, it can happen that a (short) loop of Fl_Counter callbacks can be triggered by a single (very short) mouse click on one of the Fl_Counter buttons.

I don't have a solution for point (3) which is, as it seems, platform specific (macOS). I added label "Platform: macOS" to reflect this fact although points (1) and (2) could be reproduced on Linux as well (Windows not tested).

I'll post my test program and instructions for point (3) in a follow-up comment.

**Comment by ggarra13** [comment]
@Albrecht-S On my Mac Book Pro (I9, 2019) I get exactly 3 repetitions on the callback whether I try it with 500 ms or 5000ms.  If I run it with the normal 100ms, all is fine (the callback is triggered once).

**Comment by Albrecht-S** [comment]
@ggarra13 Thanks, Gonzalo, for testing. This is another interesting data point but I don't know right now how to proceed. If we assume that Rob sees exactly two iterations (looking forward to see his results), what can we conclude? I have no idea.

Tomorrow I'll try to debug this further ...

Or maybe Manolo (@ManoloFLTK) has an idea?

**Comment by ramcdona** [comment]
@Albrecht-S I am seeing the same thing as @ggarra13 for the test application counter.cxx.  Building your sample application using the tip of master 232743c3a.  I7 MacBook Pro 2017, MacOS 12.3.1 (Monterey).

Three repeats for 500ms and 1500ms.  'Normal' behavior for 100ms.

When I build my 'real' application, I actually see somewhat erratic behavior.  Sometimes it advances one, usually two, and sometimes even three.  When it advances three, it skips one of the visual updates -- instead of 5, 6, 7, 8 I see 5, 7, 8.

**Comment by Albrecht-S** [comment]
Thanks, Rob, for your feedback. Meanwhile I made some progress although I don't have a good solution for the entire problem. At least I know now **what** is happening but I don't know yet why this happens **only on macOS**.

In short words: since your callback doesn't call `Fl::wait()` the handling of the `FL_RELEASE` event is postponed _after_ the time when your callback returns. On macOS it seems that the timeout scheduled by the click on the `Fl_Counter` button that starts the auto repeat sequence is handled **before** the pending `FL_RELEASE` event  - on Linux it seems to be the opposite which is more as expected. **_I'll look into this later when time permits._**

A "simple" ;-) way to work around this is to write at the beginning of your callback something like this (assuming `counter` is a pointer to your `Fl_Counter` widget):
```
  counter->handle(FL_RELEASE); // hack: workaround for Fl_Counter
  counter->deactivate();       // block repeating clicks
  counter->redraw();           // make changes visible
  Fl::wait(0);                 // give FLTK a chance ...
```
At the end of the callback you should reactivate the `Fl_Counter` widget:
```
  counter->activate();         // reactivate the counter widget
```
Some of this is good (FLTK) programming standard, some of this is also a little hack'ish.

What is this all about? In all FLTK programs a callback should return quickly. If this is not possible or not desired (as in your case) you should
1. call Fl::wait() from time to time to keep the UI responsive
2. deactivate the UI as far as necessary to block recursive calls of your callback.

Notes:
- Calling `handle(FL_RELEASE);` explicitly is a hack. It makes sure that the counter can't trigger the auto-repeat function.
- Calling `Fl::wait()` only once is a compromise. You should call it regularly during your calculations if possible.

In my tests this worked reliably and as expected on macOS.

In your 'real' application I recommend not to use `Fl_Counter` as you mentioned earlier. A simple `Fl_Button` wouldn't need the `FL_RELEASE` hack because it triggers the callback on release of the button anyway (unless you change this). Thus, calling `deactivate()` and `Fl::wait()` would likely suffice (maybe you need `redraw()` too). If you need status info (the "counter") then another box or the button label would be a good place to display this info.

You could also change the button label to "working (1/3) ..." while the callback executes, or something similar.

That said, AFAICT we can say now that the handling of timeouts and the long running callback w/o calling `Fl::wait()` don't work well together. It's in a gray zone of system specific behavior.

However, I'm not going to finish my investigation and if I can find a solution of the original problem (which is quite difficult) I'll let you know here.

**Comment by ManoloFLTK** [comment]
@Albrecht-S Here is, I believe, the origin of the different behaviours under macOS and X11.

Both methods `Fl_Darwin_System_Driver::wait(double time)` and `Fl_Unix_System_Driver::wait(double time)` call a function called `do_queued_events`. 
- The macOS version of this function processes 0 or 1 system event and then returns.
- The X11 version processes all queued events via this loop 
```
  while (XEventsQueued(fl_display,QueuedAfterReading)) {
    XEvent xevent;
    XNextEvent(fl_display, &xevent);
    if (fl_send_system_handlers(&xevent))
      continue;
    fl_handle(xevent);
  }
```
Consequently, under X11, the FL_RELEASE event is processed after the end of the long lasting callback() function of the test program. In contrast, under macOS, the program goes again through Fl::wait() which makes it process the repeated timeouts before it processes the FL_RELEASE event.

If the macOS version of `do_queued_events()` is modified from the present
```
  if (event != nil) {
    got_events = 1;
    [FLApplication sendEvent:event]; // will then call [NSApplication sendevent:]
  }
```
to
```
  if (event != nil) {
    got_events = 1;
    [FLApplication sendEvent:event]; // will then call [NSApplication sendevent:]
    while ( (event = [NSApp nextEventMatchingMask:NSAnyEventMask
                                        untilDate:[NSDate dateWithTimeIntervalSinceNow:0]
                                          inMode:NSDefaultRunLoopMode dequeue:YES]) != nil) {
      [FLApplication sendEvent:event]; // will then call [NSApplication sendevent:]
    }
  }
```
the test program runs identically under macOS and X11.

Now, modifying this function of the macOS-specific FLTK code deserves much caution and much analysis. I'm not sure this issue deserves it. What do you think?

**Comment by Albrecht-S** [comment]
@ManoloFLTK Thank you very much for your analysis. As you may know I'm not familiar with the macOS specific code and internals.

> modifying this function of the macOS-specific FLTK code deserves much caution and much analysis.

First of all, I think that the event handling code should behave as similar as possible on all platforms. I'm pretty sure that the Windows code would also handle all pending events (once it is at the point to handle events) and then, after the last event was handled, go further and eventually process timeouts. If this is different under macOS we should really consider unifying this.

OTOH, I also thought about checking the timer queue after every single event but that seemed too much overhead and I didn't pursue it. However, according to what you said, this is what the macOS code effectively does today.

> I'm not sure this issue deserves it. What do you think?

I'd say that this issue only revealed the different kinds of behavior we saw. Therefore I think it should be adjusted - either on macOS or on (all) other platforms. If your fix shown above works, then this is a very good starting point and it certainly explains the different behavior which is good and important to know.

That said, there's another issue with event handling and timeouts I would like to investigate further which is very closely related to the issue here: we might need **_some kind of serialization of event handling and timeouts_**.

Before we investigate this issue further we should discuss that  topic ("serialization") but for now I think this would be OT here and I will send you a private mail. Maybe this discussion will eventually solve the issue here as well.

**Comment by ManoloFLTK** [comment]
The problem should be fixed now.
And also under Wayland where it was present too.

**Comment by Albrecht-S** [comment]
@ramcdona Rob, can you confirm that the latest commit fixes your issue? If yes, the please close this issue.

FTR: The mentioned topic _"serialization of event handling and timeouts"_ will be pursued independently of this issue.

**Comment by ramcdona** [comment]
Tested with 2f52b54a0.  Seems to be working as expected.

Thanks @ManoloFLTK  and @Albrecht-S
